tardis/montecarlo/src/integrator.c

```diff
@@ -0,0 +1,223 @@
+#include <stdbool.h>
+#include <string.h>
+#include <assert.h>
+#include <stdio.h>
+#include <math.h>
+
+#include "integrator.h"
+
+#ifdef WITHOPENMP
+#include <omp.h>
+#endif
+
+#define NULEN   0
+#define LINELEN 1
+#define PLEN    2
+#define SHELLEN 3
+
+#define M_PI acos(-1.0)
+#define C_INV 3.33564e-11
+
+indexpair_t find_nu_limits_for_crossing_and_p(double nu, double p, int cr_idx, int no_of_cr_shells, double inv_ct, const double* Rs, const double* line_nu, int len)
+{
+    double blu_R, red_R, z_blu, z_red, z_cr, nu_blu, nu_red;
+    indexpair_t pair;
+
+    if (no_of_cr_shells > 1)
+    {
+        assert(Rs[0] > Rs[1]); // Decreasing order
+        
+        blu_R = get_r(cr_idx,no_of_cr_shells,Rs);
+        red_R = get_r(cr_idx+1,no_of_cr_shells,Rs);
+        z_blu = sqrt( blu_R*blu_R - p*p );
+        z_red = sqrt( red_R*red_R - p*p );
+        nu_blu = nu * (1 - get_cr_sign(cr_idx,no_of_cr_shells)*z_blu*inv_ct);
+        nu_red = nu * (1 - get_cr_sign(cr_idx,no_of_cr_shells)*z_red*inv_ct);
+    }
+    else 
+    {
+        z_cr = sqrt( Rs[cr_idx]*Rs[cr_idx] - p*p );
+        nu_blu = nu * (1 + z_cr*inv_ct);
+        nu_red = nu * (1 - z_cr*inv_ct);
+    }
+
+    pair = nu_idx_from_nu_pair(nu_blu, nu_red, line_nu, len);
+ 
+    return pair;
+}
+indexpair_t nu_idx_from_nu_pair(double nu_blu, double nu_red, const double* line_nu, int len)
```

Comment:
have a look at `line_search` from cmontecarlo.c/.h
This does the same but uses an optimized algorithm.